File: mycodes/54-SpiralMatrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        if (matrix.length == 0 || matrix[0].length == 0) return res;
        int M = matrix.length, N = matrix[0].length;
        for (int i = 0; i <= (Math.min(M, N) + 1) / 2 - 1; ++i) {
            for (int j = i; j <= N - 1 - i; ++j) res.add(matrix[i][j]);
            for (int j = i + 1; j <= M - 1 - i; ++j) res.add(matrix[j][N - 1 - i]);
            for (int j = N - 2 - i; i != M - 1 - i && j > i; --j) res.add(matrix[M - 1 - i][j]);
            for (int j = M - 1 - i; i != N - 1 - i && j > i; --j) res.add(matrix[j][i]);
        }
        return res;
    }
}

File: mycodes/54-SpiralMatrix.go
package main

func spiralOrder(matrix [][]int) []int {
  res := make([]int, 0, len(matrix)*len(matrix[0]))
  halfwidth := min((len(matrix)+1)/2, (len(matrix[0])+1)/2)
  for i := 0; i < halfwidth; i++ {
    res = append(res, getOrderOfBoarder(i,i,matrix)...)
  }
  return res
}

func getOrderOfBoarder(starti, startj int, matrix [][]int) []int {
  endj := len(matrix[starti])-startj
  endi := len(matrix)-starti
  res := make([]int, 0)
  for j := startj; j < endj; j++ {
    res = append(res, matrix[starti][j])  
  }
  for i := starti+1; i < endi-1; i++ {
    res = append(res, matrix[i][endj-1])  
  }
  for j := endj-1; j >= startj && endi-1 != starti; j-- {
    res = append(res, matrix[endi-1][j])  
  }
  for i := endi-2; i > starti && endj-1 != startj; i-- {
    res = append(res, matrix[i][startj])  
  }
  return res
}
